Use reaper process instead of shutdown hooks for testclusters

rjernst opened the pull request:
Testclusters currently provides protection from clusters living past the
life of a build by adding a shutdown hook to java. While this works in
some cases, it does not cover all cases like where the daemon is killed
with SIGKILL.

To handle these other cases, this commit replaces the shutdown hooks with
a separate process (one per build) that manages reaping external services
if gradle dies.

rjernst committed:
Use reaper process instead of shutdown hooks for testclusters

Testclusters currently provides protection from clusters living past the
life of a build by adding a shutdown hook to java. While this works in
some cases, it does not cover all cases like where the daemon is killed
with SIGKILL.

To handle these other cases, this commit replaces the shutdown hooks with
a separate process (one per build) that manages reaping external services
if gradle dies.

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst committed:
fix reaper project location

rjernst commented:
@elasticmachine run elasticsearch-ci/1 
@elasticmachine run elasticsearch-ci/bwc

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:
```diff
@@ -0,0 +1,141 @@
+            public void buildFinished(BuildResult result) {
+                if (reaperProcess != null) {
+                    try {
+                        reaperProcess.getOutputStream().close();
```
As mentioned below, I think this might cause funkiness in some scenarios as we are effectively closing the Gradle process inputstream here. How does this even work for long-lived daemons? Does this get reinitialized on every build? On a non-daemon build what happens? Probably best to control this ourselves rather than rely on implicit behavior.

rjernst replied:
This is the key of how the reaper works. It waits on stdin for a read. That call will not complete until the pipe with the parent process (here) is broken. In this case, we are explicitly closing since we had a normally completing build (gradle did not die). It is my understanding Plugin instances are not reused across builds within the same daemon. So whether or not the daemon is long lived, this should signal the reaper at the end of the build to shutdown. What implicit behavior does it look like we are relying on?

mark-vieira replied:
That comment was based on my assumption that the `ProcessBuilder` passed along the parent's input directly, rather than via a pipe so I think we are ok here in that we will never call `close()` on the same stream twice.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:
```diff
@@ -0,0 +1,141 @@
+        ensureReaperStarted();
+
+        try {
+            Files.writeString(baseDirectory.resolve(serviceId), String.join(" ", command));
```
Should we be including a newline here somewhere?

rjernst replied:
No need. While the reaper does handle multiple lines, we don't actually use that. It will end up reading just the entire content into a single line from the file.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:
```diff
@@ -0,0 +1,141 @@
+    private volatile Process reaperProcess;
+
+    @Override
+    public void apply(Project project) {
```
We need to add a check here to ensure this plugin is only ever applied to the root project.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java:
```diff
@@ -56,15 +57,18 @@
         project.getPlugins().apply(DistributionDownloadPlugin.class);
 
+        project.getRootProject().getPluginManager().apply(ReaperPlugin.class);
+        reaper = project.getRootProject().getPlugins().getPlugin(ReaperPlugin.class);
```
Grabbing the plugin instance itself seems quite wrong. Can we put the bits we need here in a project extension applied by the plugin and use that instead?

rjernst replied:
I was trying to keep the logic in the plugin, as was suggested previously. :) There isn't any actual configuration, which is why I didn't add an extension. I think we would want the state to remain in the plugin (the process instantiation) so what would the extension be other than a wrapper around its own ref to the plugin?

mark-vieira replied:
Yeah, there's just something about passing around plugin instances that seems nasty to me. In fact, I wouldn't be surprised if these APIs are deprecated in the future. The `getPlugin()` method here is really meant for getting at the plugin convention objects (since replaced by extensions). What we really want is a singleton "service" but since Gradle's DI stuff isn't public we go with a workaround here. I think using an extension for the purpose is the best next option rather than passing references to a plugin instance. We are battling between the ideas the "plugins should be smart, extensions should be dumb" and "only Gradle should handle plugin instances". In the end I think the latter takes precedence over the former as I think these "service objects" are best implemented as extensions and not plugins.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:
```diff
@@ -0,0 +1,141 @@
+    public void registerPid(String serviceId, long pid) {
+        String[] killPidCommand = OS.<String[]>conditional()
+            .onWindows(() -> new String[]{"Taskill", "/F", "/PID", String.valueOf(pid)})
+            .onUnix(() -> new String[]{"kill", "-9", String.valueOf(pid)}).supply();
```
nit: can we put the call to `.supply()` on a newline here?

mark-vieira commented on buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java:
```diff
@@ -0,0 +1,125 @@
+
+    public static void main(String[] args) throws Exception {
+        if (args.length != 1) {
+            System.err.println("Expected one argument.\nUsage: java -cp reaper.jar Reaper <DIR_OF_REAPING_COMMANDS>");
```
FYI, this usage line won't actually work because we need to provide the fully qualified name of the class. Or we could just do as I mentioned in another comment and make this an executable jar and use `java -jar`.

mark-vieira commented on buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java:
```diff
@@ -0,0 +1,125 @@
+ */
+public class Reaper implements Closeable {
+
+    Path inputDir;
```
Both these instance variables can, and probably should, be `private`.

mark-vieira commented on buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    private void reap() {
+        try {
```
I think we might want to make this exception handling a bit more granular. For example, if we blow up trying to kill one process the check in `close()` will mean we don't clean up any of these command files. Later if we try to run this again calling `kill` against an already killed pid will fail, which will mean `failed == false` and this cycle continues. 

To completely get around this we'll probably have to delete the command files as we successfully execute them such that if we blow up mid-way we can pick up where we left off next time. We still might want to do per-command exception handling though such that we can clean up as much as we can. We don't want to keep any more stuff dangling than we have to for whatever reason.

rjernst replied:
We read the files up front, before any processing, specifically so that we don't rely on the state of the directory changing. There isn't any possibility for a failure cycle.

While it is possible we could blow up when trying to log a failure of one of the commands (getting an ioexception when writing to the log), we are pretty hosed at that point, so I'm not sure special handling is really worthwhile, since some investigation would need to be done anyways. There is no "next time" with this, since the directory is per gradle daemon. The only way the reaper should actually do work is if the daemon died (since we should be unregistering everything in the normal case).

mark-vieira replied:
> We read the files up front, before any processing, specifically so that we don't rely on the state of the directory changing. 

We only read the _list_ of files. What I'm saying is, what if I have 5 processes to kill and something happens on say, the third one? Right now we would immediately stop, not even attempting to kill the final 2, and mark reaping as having failed. On `close()` we would then not delete _any_ of the pid files meaning the next time we go to reap we would immediately fail again on the first file because the process _had_ been reaped, but we haven't marked it as such.

mark-vieira replied:
> The only way the reaper should actually do work is if the daemon died (since we should be unregistering everything in the normal case).

Not necessarily, we could fail to stop these processes gracefully for any number of reasons that doesn't necessarily involve the Gradle daemon dying.

rjernst replied:
> we could fail to stop these processes gracefully for any number of reasons that doesn't necessarily involve the Gradle daemon dying.

This shoudln't be possible. Testclusters falls back to using SIGKILL if SIGTERM doesn't work, and if SIGKILL somehow fails, it fails the build.

>  meaning the next time we go to reap

There is no next time. The reaper is per build, and it only does work at the end of the build, when gradle died.

mark-vieira replied:
The directory with the files is based on the daemon pid though. Isn't it possible the next build with the same daemon will run a reaper pointed at the same directory?

rjernst replied:
We delete the directory completely, and in this case the *only* thing that could fail is deleting our log file just before deleting the dir (since those are the only two files when we exited normally). But if that happens we have bigger problems (if we can't delete, then we likely have permission problems, which will persistent regardless). So the next build with the same daemon will create a fresh copy of the dir.

mark-vieira replied:
We only delete the directory if we don't encounter any failures. As I described above, if any of our reaping commands fail, we don't clean up any of this stuff. There's any number of reasons why this might happen (processes crash, are killed manually outside Gradle, etc). In such a case so long as that daemon lives we'll just keep trying and trying, and failing and failing. At the very least we should delete these files as we go, such that when we do successfully reap a process we immediately clean up the pid file.

rjernst replied:
> As I described above, if any of our reaping commands fail, we don't clean up any of this stuff. 

If there are any reaping commands to run, then gradle died. We should be unregistering all commands from the reaper in any build that completes, whether a success or failure.

I have made the delete happen on successful execution of the command. If we end up having problems for some reason, we can randomize this directory to make it unique per build instance instead of daemon, but I think that can be a followup, and only if we run into issues.

mark-vieira replied:
> If there are any reaping commands to run, then gradle died. We should be unregistering all commands from the reaper in any build that completes, whether a success or failure.

Ah, understood.

I wonder if we are going to need anything to periodically clean these directories up? It's very possible that we later launch a daemon with the same pid, which might in turn have stale stuff in there.

rjernst replied:
Since the directory is deleted in the normal case, I think leaving it to the user to cleanup if old directories exist is fine. Adding random element to the dir name as I described above would stop a future daemon from reusing a potentially failed daemon's dir.

mark-vieira replied:
Yeah, my only concern is that if we ran into a situation where we were using an old daemon dir it would be hard for a user to figure that out. I'd say the scenario of reusing an old pid that _also_ failed to reap processes is fairly low.

Going forward the preferred method would be some kind of protocol that we communicate with the reaper with over stdin. That way the reaper process itself holds all this state and we can ditch all this filesystem stuff. We can tackle that another day though.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:
```diff
@@ -0,0 +1,141 @@
+        this.project = project;
+        this.baseDirectory = project.getRootDir().toPath().resolve(".gradle")
+            .resolve("reaper").resolve("build-" + ProcessHandle.current().pid());
+        project.getGradle().addListener(new BuildListener() {
```
You can avoid a bunch of this anonymous class boilerplate by either a) create a new anonymous instance of `BuildAdapter` which already implements `BuildListener` methods so you can just override the one you care about or b) call `Gradle.buildFinished()` and provide your listener as a lambda.

rjernst replied:
Thanks for the suggestion! I was not aware of either of these. I switched to a lambda with Gradle.buildFinished

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:
```diff
@@ -0,0 +1,141 @@
+        if (reaperProcess == null) {
+            try {
+                // copy the reaper jar
+                Path jarPath = project.getBuildDir().toPath().resolve("reaper").resolve("reaper.jar");
```
We should look at ways of caching this. As far as I can tell we copy this jar on every build regardless of if the file exists. That said, we can't just check if the file exists as that doesn't mean it's up to date so we'll have to implement our own caching/up-to-date mechanism here.

rjernst replied:
Is it that big a deal? If we are starting the reaper, we are doing much more heavy io from eg starting an elasticsearch cluster, which involves tons of copying. The simplicity of always copying seems nice here, at least for now until it is a bottleneck? Remember we only do this in the case we are starting the reaper, which happens when we are actually launching our first elasticsearch node in testclusters.

mark-vieira replied:
_Probably_ not, but this is blocking I/O in a `synchronized` method which means we can't proceed to start anymore nodes until this bit is done. We are probably talking milliseconds here but it's something to think about.

rjernst replied:
Milliseconds for the very first cluster started. Given that the vast majority of our test clusters are single node, I don't think parallelism is hurt, and these clusters are very heavyweight. Can we defer this until it is a problem?

mark-vieira replied:
Agreed. This isn't a hotspot in our build.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:
```diff
@@ -0,0 +1,141 @@
+
+                // start the reaper
+                ProcessBuilder builder = new ProcessBuilder(
+                    "java", "-cp", jarPath.toString(), "org.elasticsearch.gradle.reaper.Reaper", baseDirectory.toString());
```
We should also use some pretty conservative JVM memory settings here instead of going with platform defaults. We know this thing is lightweight so no need to give 1G of max heap or whatever the default is.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:
```diff
@@ -0,0 +1,141 @@
+                File logFile = baseDirectory.resolve("reaper.log").toFile();
+                builder.redirectOutput(logFile);
+                builder.redirectError(logFile);
+                reaperProcess = builder.start();
```
One missing bit in terms of error handling here is that we won't get any kind of exception if we simply fail to launch to reaper itself since `start()` is effectively async.

We should probably check for `isAlive()` at times. Perhaps whenever we call one of the `register()` methods, and also in `buildFinished()`. That way if for some reason the process shit the bed when we didn't expect we'd _eventually_ be notified.

mark-vieira reviewed (changes requested):
I've highlighted some things I think need to be addressed. Otherwise I really like this concept. It seems much more robust than relying on Gradle or JVM hooks that may or may not fire in certain scenarios.

rjernst commented on buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java:
```diff
@@ -0,0 +1,125 @@
+    public void close() {
+        if (failed == false) {
+            try {
+                for (Path toDelete : Files.walk(inputDir).sorted(Comparator.reverseOrder()).collect(Collectors.toList())) {
```
That won't work, since we need the root directory to be deleted last, otherwise it will fail. Thus we need the reverse sort to guarantee the root comes last.

mark-vieira replied:
Can't we do `.forEach()` after the sort?

rjernst replied:
Nope, because Files.delete throws IOException, and there is no functional interface that throws exceptions (we have them in ES, but not in our build code). So we would have to add a try/catch within the lambda, which I thought would make this more difficult to understand, and we would still have to wrap it in an unchecked exception, which would just pop back out to the catch here.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:
```diff
@@ -0,0 +1,141 @@
+                Files.createDirectories(baseDirectory);
+
+                // start the reaper
+                ProcessBuilder builder = new ProcessBuilder(
```
The default is PIPE (see https://docs.oracle.com/javase/7/docs/api/java/lang/ProcessBuilder.html#redirectInput()), and we rely on it (see how we close the pipe in buildFinished for exiting normally). I'm happy to make it explicit though.

mark-vieira replied:
Ah ok. So the inputstream we are closing isnt _actually_ the parent process? But I think we still have the issue for non-daemons in the stdin may actually be the terminal. We should test this. Also ensure that we aren't launching a single-use daemon either but are running the build in "embedded" mode.

rjernst replied:
The input we close is the input stream of the child process, not of gradle. It is just a pipe; it's a completely different file descriptor from stdin of gradle.

rjernst commented on buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java:
```diff
@@ -0,0 +1,125 @@
+
+    private void logFailure(String message, Exception e) {
+        System.err.println(message);
+        e.printStackTrace(System.err);
```
Oops, I lost a null check in a refactoring. Will fix.

rjernst committed:
Merge branch 'master' into distro_tests17

rjernst committed:
address feedback

rjernst commented:
@mark-vieira I believe I have addressed all your feedback in https://github.com/elastic/elasticsearch/pull/44927/commits/492697961ad133f9907d34a7c77698b24f1b88a0

rjernst commented:
@elasticmachine run elasticsearch-ci/1

rjernst commented:
@elasticmachine run elasticsearch-ci/1

alpar-t reviewed (changes requested):
Thanks for taking this on! I feel guilty for not getting to it as promised. 
It's a great improvement and one we can use for test fixtures too not just test clusters !

rjernst committed:
Merge branch 'master' into distro_tests17

rjernst commented on buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java:
```diff
@@ -0,0 +1,121 @@
+            System.err.println("Expected one argument.\nUsage: java -jar reaper.jar <DIR_OF_REAPING_COMMANDS>");
+            System.exit(1);
+        }
+        Path inputDir = Paths.get(args[0]);
```
I think that is overkill. We don't do this for any other directories in the build.

atorok replied:
Maybe at least create the directory with such restricted permissions instead of relying on the default mask.  
This directory is different because we execute everything that gets written there, so it transforms the ability to write a file to that location into the ability to execute as the user running the daemon. 
Suppose an attacker gains a way to create files, this gives them the ability to execute them as a different user and do thinks like steal secrets available to Jenkins.

rjernst replied:
That is an interesting case I had not thought of. However, Windows doesn't have permissions like this to set, so we wouldn't be covered in all cases. Additionally, the user that an attacker would likely be writing files as would the same user that gradle runs with, ie jenkins.

I think the best thing long term is to create a wire format to pass information to the reaper over stdin. However, I would like to do that as a followup because this currently works, and the case you described is an extreme edge case.

mark-vieira replied:
Fun fact, the Gradle daemon itself isn't secured in any way. Anyone that communicate with it over a socket can issue arbitrary build actions that will run as the daemon user.

I think the approach I mentioned above for using stdin as the communications channel for which we tell the reaper about processes to kill is the better solution here vs relying on filesystem permissions. No one could hijack that as the reaper would effectively be "owned" by the daemon and only it could issue instructions. We can defer that for a later date though I think.

rjernst commented on buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java:
```diff
@@ -0,0 +1,121 @@
+
+    private void reap() {
+        try {
+            final List<Path> inputFiles = Files.list(inputDir)
```
I explicitly want to collect all these before we start hte loop. Using an active stream over the list will mean we can pick up changes to the files. This was more important to avoid in a first iteration when I dynamically added the log file, but I do like iterating over something that can't change so I will keep this as is.

atorok replied:
What I meant is that per the docs  `Files.list` returns a `Stream ` that  needs to be closed and we are not doing that here. You can still collect it to a list first, just make sure to close the stream.

rjernst replied:
I don't think it is necessary. From the Stream docs:

```
Streams have a BaseStream.close() method and implement AutoCloseable, but nearly all stream instances do not actually need to be closed after use. Generally, only streams whose source is an IO channel (such as those returned by Files.lines(Path, Charset)) will require closing. Most streams are backed by collections, arrays, or generating functions, which require no special resource management.
```

Doing a filesystem operation here is not actually doing IO (it's only filesystem metadata).

atorok replied:
I was looking at [Files.list](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#list-java.nio.file.Path-) and [DirectoryStream](https://docs.oracle.com/javase/8/docs/api/java/nio/file/DirectoryStream.html) specifically ```Failure to close the stream may result in a resource leak.```

rjernst replied:
TIL (!). Thanks @atorok. I pushed c057be6

rjernst commented on buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java:
```diff
@@ -0,0 +1,121 @@
+                System.out.println("Process file: " + inputFile);
+                for (String line : Files.readAllLines(inputFile)) {
+                    System.out.println("Running command: " + line);
+                    String[] command = line.split(" ");
```
I thought about doing explicit commands, but this makes the reaper much less useful as a general purpose tool, because it must be modified every time there is a new usecase.

> Not dealing with platform specifics hear will mean that he complexities of doing so will leek to all the users of this functionality.

Platforms specifics are a reality. The user of the reaper knows better how to deal with shutting down a service on the current platform than the reaper could.

I chose to make this simple for now. We can always change things like making spaces work, or even allowing multiple lines (eg for trying a SIGTERM followed by SIGKILL if that doesn't work after some timeout), but I don't see a concrete need for such complexity yet.

atorok replied:
I was on the fence with it too, I'm fine to leave it as it is for now.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperService.java:
```diff
@@ -0,0 +1,136 @@
+                    "-Xms4m", "-Xmx16m", // no need for a big heap, just need to read some files and execute
+                    "-jar", jarPath.toString(),
+                    inputDir.toString());
+                logger.lifecycle("Launching reaper: " + String.join(" ", builder.command()));
```
Doh! This was leftover from debugging. I will switch back to info.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperService.java:
```diff
@@ -0,0 +1,136 @@
+            .onWindows(() -> new String[]{"Taskill", "/F", "/PID", String.valueOf(pid)})
+            .onUnix(() -> new String[]{"kill", "-9", String.valueOf(pid)})
+            .supply();
+        registerCommand(serviceId, killPidCommand);
```
I disagree, when removing the variable makes a single statement more complex. I dislike passing a single multiline argument to a function, and prefer to have function calls be single line if possible, thus the local here.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperService.java:
```diff
@@ -0,0 +1,136 @@
+    /**
+     * Register a system command that will be run by the reaper.
+     */
+    public void registerCommand(String serviceId, String... command) {
```
This is public because I will need it in a future PR for registering vagrant halt commands. I only added the pid method here as a convenience because I thought it would be a fairly common thing to do. This method is the real intended api, the pid method is just sugar.

mark-vieira replied:
I concur. Let's make this `private` for now, and if future processes need some other mechanism to be reaped beyond `kill` we can bake that into the plugin in a safe and robust way.

rjernst replied:
What would making this private (and thus requiring all types of services the reaper can work on be statically known) achieve? I see this as equivalent to if we made the java's that could be used with javaexec static. This is the API we are providing for the reaper, and consumers (other plugins) use it to register system commands they would run themselves. As I mentioned above, I will be using this in a followup PR for ensuring vagrant is shut down. I don't see any reason the reaper knowing about vagrant would make this any more robust. A command would still have to be generated, somewhere, whether here, in the reaper itself, or by the plugin using the reaper. In the last case, the logic for how to deal with vagrant stays with the plugin that started vagrant, which I think is much better than having some magic reaper that happens to know how to shutdown a handful of services.

mark-vieira replied:
I don't have strong feelings here. If the API is "here's a general way to run arbitrary commands" then that's fine. As alpar said, we just have to ensure that usages are good about ensuring these things are quick and relatively robust.

atorok replied:
I don't have strong feelings either, it just seems a bit too broad.
E.x. for vagrant having something like a public `registerLinuxCommand` that asserts that we are running on Linux might make more sense so we fail fast if accidentally registering vagrant shutdown commands   on Windows.
I'm also fine to leave it like ti is.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/ReaperService.java:
```diff
@@ -0,0 +1,136 @@
+                // start the reaper
+                ProcessBuilder builder = new ProcessBuilder(
+                    Jvm.current().getJavaExecutable().toString(), // same jvm as gradle
+                    "-Xms4m", "-Xmx16m", // no need for a big heap, just need to read some files and execute
```
Does this really only use 16m of heap? Damn, Java's gotten lighter over the years.

rjernst replied:
I can get a basic java program to work with 8m heap, but I thought 16 would be conservative. :)

rjernst committed:
more feedback

mark-vieira reviewed (approved):
Changes LGTM 👍

rjernst commented:
@atorok I believe I have address all of your comments.

rjernst commented:
@elasticmachine run elasticsearch-ci/1

rjernst committed:
use try with resources

alpar-t reviewed (approved):
LGTM

rjernst committed:
Merge branch 'master' into distro_tests17
